### PR TITLE
CMA ES for hyperparameter search and plugin architecture

### PR DESCRIPTION
**Pull request by jasonzliang**
This adds in CMA-ES for hyperparameter search. It also creates a new folder called optimizer_plugins,
where you can put your custom plugins.
The custom plugins must following the following api, there must be a object with class name "Optimizer" and implement the following public methods below:

```
while not optimizer.stop():
                solutions = optimizer.ask()
                fitnesses = evaluate(solutions
                optimizer.tell(solutions, fitnesses)

```
There are still some missing features to be added in future:

Optimization of numpy arrays
Unbounded optimization
Here is an example of how to run it:

studio run -opt=cmaes -hp=momentum=0.01:0.01:0.05 -hp=lr=0.01:0.01:0.05 studio_test.py
studio_test.py source code:

```
import sys

lr = 0.01
momentum = 0.01

print "fitness: %s" % (lr + momentum)
```

### CODE REVIEW

**Comment by pzhokhov** [comment]
Looks good to me, but could you add some tests ? I know we don't have them running as a part of the build now, but I'd like to get there at some point. 
So far the tests live in tests/ folder; and I manually run them via `python $(which nosetests) --processes=4 --process-timeout=900`

File: studio/runner.py [hunk 1]
```diff
@@ -278,6 +340,7 @@ def main(args=sys.argv):
     else:
         if queue_name == 'local':
             queue = LocalQueue()
+            queue.clean()
```
Comment (pzhokhov) [inline studio/runner.py]: Like Karl pointed out, I think we should not clean the local queue before running the experiments. We can add a command that does it to the CLI; but generally the experiment ends up in a local queue unprocessed is a sign that something went wrong in the runner (i.e. when local_worker works correctly, this should not be happening), so I'd rather fix the problems there.

Comment (jasonzliang) [reply studio/runner.py]: There is a bug where the code crashes if the folders in ~/studioml/experiments remain, but the corresponding file in the local queue is deleted. What would you suggest to do to fix this issue?

Comment (pzhokhov) [reply studio/runner.py]: Could you provide an example (name of the branch + what command are you running)? In principle, data in ~/.studioml/experiments is just cache of experiment data; it should not be deleted when the queue is cleaned.

Comment (jasonzliang) [reply studio/runner.py]: we will leave in queue.clean() for now, will address this in a future PR

**Pull request merged**